dali/base/dacoven.cpp

```diff
@@ -155,6 +155,78 @@ class CDaliUidAllocator: public CInterface
 
 };
 
+static void checkDaliVersionInfo(ICommunicator *comm, CDaliVersion &serverVersion, CDaliVersion &minClientVersion)
+{
+    CMessageBuffer mb;
+    mb.append(MCR_GET_VERSION_INFO);
+    mb.append(ClientVersion);
+    mb.append(MinServerVersion);
+    if (!comm->sendRecv(mb, RANK_RANDOM, MPTAG_DALI_COVEN_REQUEST, VERSION_REQUEST_TIMEOUT))    
+        throw makeStringException(-1, "failed retrieving version information from server, legacy server?");
+    if (!mb.length())
+        throw makeStringException(-1, "Failed to receive server information (probably communicating to legacy server)");
+    StringAttr serverVersionStr, minClientVersionStr;
+    mb.read(serverVersionStr);
+    serverVersion.set(serverVersionStr), 
+    mb.read(minClientVersionStr);
+
+    CDaliVersion clientV(ClientVersion);
+    minClientVersion.set(minClientVersionStr);
+    if (clientV.compare(minClientVersion) < 0)
+    {
+        StringBuffer s("Client version ");
+        s.append(ClientVersion).append(", server requires minimum client version ").append(minClientVersionStr);
+        throw createClientException(DCERR_version_incompatibility, s.str());
+    }
+    CDaliVersion minServerV(MinServerVersion);
+    if (serverVersion.compare(minServerV) < 0)
+    {
+        StringBuffer s("Server version ");
+        s.append(serverVersionStr).append(", client requires minimum server version ").append(MinServerVersion);
+        throw createClientException(DCERR_version_incompatibility, s.str());
+    }
+}
+
+static CTimeLimitedCache<std::string, std::pair<std::string, std::string>> *foreignDaliVersionCache = nullptr;
+MODULE_INIT(INIT_PRIORITY_STANDARD)
+{
+    foreignDaliVersionCache = new CTimeLimitedCache<std::string, std::pair<std::string, std::string>>(1000*60*5); // 5mins
```

Comment:
Did you consider mapping to a std::pair<CDaliVersion, CDaliVersion>?  It would simplify some of the code a little.

Reply:
agree, have changed.